packages/firebase_messaging/android/src/main/java/io/flutter/plugins/firebasemessaging/FlutterFirebaseMessagingService.java

```diff
@@ -42,4 +128,148 @@ public void onNewToken(String token) {
+      }
+      sBackgroundMessageQueue.clear();
+    }
+  }
+
+  public static void setBackgroundChannel(MethodChannel channel) {
+    sBackgroundChannel = channel;
+  }
+
+  public static void setPluginRegistrant(PluginRegistry.PluginRegistrantCallback callback) {
+    sPluginRegistrantCallback = callback;
+  }
+
+  public static void setBackgroundMessageHandle(Context context, Long handle) {
+    sBackgroundMessageHandle = handle;
+
+    // Store background message handle in shared preferences so it can be retrieved
+    // by other application instances.
+    SharedPreferences prefs = context.getSharedPreferences(SHARED_PREFERENCES_KEY, 0);
+    prefs.edit().putLong(BACKGROUND_MESSAGE_CALLBACK_HANDLE_KEY, handle).apply();
+  }
+
+  public static void setBackgroundSetupHandle(Context context, long callbackHandle) {
+    // Store background setup handle in shared preferences so it can be retrieved
+    // by other application instances.
+    SharedPreferences prefs = context.getSharedPreferences(SHARED_PREFERENCES_KEY, 0);
+    prefs.edit().putLong(BACKGROUND_SETUP_CALLBACK_HANDLE_KEY, callbackHandle).apply();
+  }
+
+  public static Long getOnMessageCallbackHandle(Context context) {
+    return context
+        .getSharedPreferences(SHARED_PREFERENCES_KEY, 0)
+        .getLong(BACKGROUND_MESSAGE_CALLBACK_HANDLE_KEY, 0);
+  }
+
+  private static void executeDartCallbackInBackgroundIsolate(
+      Context context, RemoteMessage remoteMessage, final CountDownLatch latch) {
+    if (sBackgroundChannel == null) {
+      throw new RuntimeException(
+          "setBackgroundChannel was not called before messages came in, exiting.");
+    }
+
+    // If another thread is waiting, then wake that thread when the callback returns a result.
+    MethodChannel.Result result = null;
+    if (latch != null) {
+      result =
+          new MethodChannel.Result() {
+            @Override
+            public void success(Object result) {
+              latch.countDown();
+            }
+
+            @Override
+            public void error(String errorCode, String errorMessage, Object errorDetails) {
+              latch.countDown();
+            }
+
+            @Override
+            public void notImplemented() {
+              latch.countDown();
+            }
+          };
+    }
+
+    Map<String, Object> args = new HashMap<>();
+    Map<String, Object> messageData = new HashMap<>();
+    if (sBackgroundMessageHandle == null) {
+      sBackgroundMessageHandle = getOnMessageCallbackHandle(context);
+    }
+    args.put("handle", sBackgroundMessageHandle);
+
+    if (remoteMessage.getData() != null) {
+      messageData.put("data", remoteMessage.getData());
+    }
+    if (remoteMessage.getNotification() != null) {
+      messageData.put("notification", remoteMessage.getNotification());
+    }
+
+    args.put("message", messageData);
+
+    sBackgroundChannel.invokeMethod("", args, result);
```

Comment:
It strange to use an empty method name here

Reply:
This was copied from the alarm manager plugin where the background channel only has a single method call so no name was provided. I'm fine with adding a name anyway so added a method name `handleBackgroundMessage`.